cyber_source/download_payment_batch_detail_report.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+require 'cybersource_rest_client'
+require 'folio_client'
+require 'csv'
+require 'date'
+require 'json'
+
+# class to download the PaymentbatchDetailReport from CyberSource
+class DownloadPaymentBatchDetailReport
+  attr_accessor :date
+
+  CONFIGURATION_DICTIONARY = {
+    merchantID: 'wfgsulair',
+    runEnvironment: "cybersource.environment.#{ENV.fetch('CYBS_ENV', 'production')}",
+    timeout: 3000,
+    authenticationType: 'http_signature',
+    enableLog: true,
+    merchantKeyId: ENV.fetch('MERCHANT_KEY_ID', nil),
+    merchantsecretKey: ENV.fetch('MERCHANT_SECRET_KEY', nil),
+    logDirectory: 'harvestlog',
+    logFilename: 'cybs'
+  }.freeze
+
+  def initialize(date)
+    # Date format: e.g. '2001-02-03'
+    @date = date ? Date.parse(date) : Date.today
+  end
+
+  def year_or_prev(date)
+    date.month == 1 ? date.prev_year.year : date.year
+  end
+
+  def main
+    credits = []
+
+    first_day = Date.new(year_or_prev(@date), @date.prev_month.month, 1)
+    last_day = Date.new(year_or_prev(@date), @date.prev_month.month, -1)
+
+    (first_day..last_day).each do |date|
+      report_date = date.strftime('%Y-%m-%d')
+      begin
+        data, status_code, headers = download_report(report_date)
+        puts report_date, status_code, headers, data
+      rescue StandardError => e
+        puts report_date
+        puts e.message
+        next
+      end
+      next unless data
+
+      data.each_line.with_index do |batch_detail, index|
+        if index > 1
+          user_id = batch_detail.split(',')[4]
+          paid = batch_detail.split(',')[8]
+          paydate = batch_detail.split(',')[12]&.chomp
```

Comment:
Is paydate in UTC? I think we need to account for folio storing the date in UTC when reconciling folio account createdDate with the batch_detail paydate.

Reply:
Cybersource reports the payment time in UTC (Note the 'Z' in the timestamp in the spec/fixtures file)